torch/csrc/jit/runtime/script_profile.cpp

```diff
@@ -0,0 +1,124 @@
+#include <torch/csrc/jit/runtime/script_profile.h>
+
+#include <atomic>
+#include <chrono>
+#include <mutex>
+#include <unordered_set>
+
+#include <c10/util/Exception.h>
+#include <torch/csrc/jit/api/function_impl.h>
+
+namespace torch {
+namespace jit {
+
+namespace {
+
+class ProfilesRegistry {
+ public:
+  bool empty() {
+    return empty_.load(std::memory_order_relaxed);
+  }
+  void addProfile(ScriptProfile& p) {
+    std::lock_guard<std::mutex> g(mutex_);
+    enabledProfiles_.emplace(&p);
+    empty_.store(false, std::memory_order_relaxed);
+  }
+  void removeProfile(ScriptProfile& p) {
+    std::lock_guard<std::mutex> g(mutex_);
+    enabledProfiles_.erase(&p);
+    if (enabledProfiles_.empty()) {
+      empty_.store(true, std::memory_order_relaxed);
+    }
+  }
+  void send(profiling::Datapoint datapoint) {
+    std::lock_guard<std::mutex> g(mutex_);
+    for (auto* p : enabledProfiles_) {
+      p->addDatapoint(datapoint);
+    }
+  }
+
+ private:
+  std::atomic<bool> empty_{true};
+  std::mutex mutex_;
+  std::unordered_set<ScriptProfile*> enabledProfiles_;
+};
+
+ProfilesRegistry& getProfilesRegistry() {
+  static auto& registry = *new ProfilesRegistry{};
+  return registry;
```

Comment:
Do not forget to address the clang-tidy errors :)

Reply:
I think clang-tidy is warning about a separate issue, which is exposing this static variable to be globally accessible. (and I don't know how to fix this quickly and elegantly, few options: 1. encapsulate this in a LeakySingleton class, like folly (https://github.com/facebook/folly/blob/master/folly/Singleton.h#L720), 2. bypass lint error 3. expose this in a private static class function, and access it with friend functions. probably 3. feels better)

re `new`ing the object, returning static variable by reference in theory isn't inherently safer, because something could potentially call into the singleton after its destruction, in other words, the leaking is intentional. Since this will be used in Python, things can be different, but still it's always better to be cautious with global destruction order. 

If we want to save one allocation here, we can implement something like folly::Indestructible (https://github.com/facebook/folly/blob/master/folly/Indestructible.h) but that requires separate effort and isn't so different using `new` essentially. 

We can also talk about this during sync up :)